Add cache mechanism to sherpa tts

mah92 opened the pull request:
Delay is a major drawback of sherpa tts on android phones, which makes it unusable for average blind people using average phones. By caching the most frequent tts requests, sherpa is made as fast as lightning on most of the time, using much less cpu and battery. Some blind individials have tested this upgrade and have been happy :)

Considered guide-lines:
- App does not cache user texts due to security and privacy concerns. Just a hash is saved from the texts.
- Usage statistics of texts(hashes) are saved to a list, and stored periodically(every 10 minutes). If cache space is filled, wav files that are less used are removed.
- Setting amount of cache and the ability to remove cached wav files are added in the MainActivity
- Present jni functions are tried to remain intact due to possible compatibility considerations.
- When changing speed, cache could be cleared automatically to update the speed of cached media as well, not implemented yet as there is a clear button on MainActivity.kt.
- Code is scanned for possible bugs with deepseek AI.

Commit:
integrated caching feature into last commit

mah92 committed:
Merge branch 'master' into master

Commit:
Speed up least repeated txt search

csukuangfj commented:
Thank you for your contribution!

Will review it.

csukuangfj commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+    std::ifstream file(file_path, std::ios::binary | std::ios::ate);
+    if (file.is_open()) {
+      used_cache_size_bytes_ -= file.tellg();
+      file.close();
```
```suggestion
```

mah92 replied:
Kept file.close() here as I want to remove the file before desctructor of the file is called.

csukuangfj commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+      std::ifstream file(file_path, std::ios::binary | std::ios::ate);
+      if (file.is_open()) {
+        used_cache_size_bytes_ += file.tellg();
+        file.close();
```
```suggestion
```

mah92 replied:
done

csukuangfj commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+      SHERPA_ONNX_LOGE("Failed to read cached file: %s", file_path.c_str());
+    }
+  } else {
+    // SHERPA_ONNX_LOGE("Cached audio not found for text hash: %s", text_hash.c_str());
```
Please remove unused code and `else`.

mah92 replied:
done

csukuangfj commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+  }
+
+  // Ensure the text_hash exists in the map before incrementing the count
+  if (repeat_counts_.find(text_hash) == repeat_counts_.end()) {
+    repeat_counts_[text_hash] = 0; // Initialize if it doesn't exist
+  }
+  repeat_counts_[text_hash]++; // Increment the repeat count
```
If the file does not exist, shall we skip updating the count?

mah92 replied:
You are correct. I wrongly incremented the count in both in the GetWavFile and AddWavFile functions. Now incrementing is done only in the GetWavFile.

csukuangfj commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+  // SHERPA_ONNX_LOGE("Removed all wav files!");
+
+  // Reset the total cache size to 0
+  used_cache_size_bytes_ = 0;
```
Do you also need to update `repeat_counts_` and `cache_vector_`?

mah92 replied:
You are correct. Thank you for your attention to the details.

csukuangfj commented on sherpa-onnx/csrc/CMakeLists.txt:
```diff
@@ -167,6 +167,7 @@ if(SHERPA_ONNX_ENABLE_TTS)
     offline-tts-vits-model.cc
     offline-tts.cc
     piper-phonemize-lexicon.cc
+    offline-tts-cache-mechanism.cc
```
please sort the filenames alphabetically.

mah92 replied:
corrected

csukuangfj commented on sherpa-onnx/csrc/offline-tts.h:
```diff
@@ -12,6 +12,7 @@
 
 #include "sherpa-onnx/csrc/offline-tts-model-config.h"
 #include "sherpa-onnx/csrc/parse-options.h"
+#include "sherpa-onnx/csrc/offline-tts-cache-mechanism.h"
```
Please sort the includes alphabetically.

mah92 replied:
corrected

csukuangfj commented on sherpa-onnx/csrc/offline-tts.h:
```diff
@@ -32,6 +33,9 @@ struct OfflineTtsConfig {
   int32_t max_num_sentences = 1;
 
+  // Path to cache_directory
+  std::string cache_dir;
```
Please create a new config for cache.

mah92 replied:
What do you mean exactly?

csukuangfj replied:
For example, OfflineTts has a config called OfflineTtsConfig.

You can create a struct called CacheMechanismConfig and put cache_dir in it. In this way, you don't need to add a new field to OfflineTtsConfig. You also don't need to add new methods to OfflineTts. (Note you need to wrap CacheMechanism and its config to Kotlin via jni)

Remember that sherpa-onnx currently supports 12 programming languages. If we change the config in C++, we also need to update APIs of other programming languages.

You can create a smart pointer of CacheMechanism and pass it to the constructor of OfflineTts. 

If later we need to extend the CacheMechanism, we can leave OfflineTts and OfflineTtsConfig untouched. We only need to change CacheMechanism or/and CacheMechanismConfig.

By the way, since the CacheMechanism is related to tts, can you rename it to OfflineTtsCacheMechanism?

In general, the filename of a class contains the class name, though there are dashes in the filename but not in the class name

Methods of CacheMechanism  don't need to be repeated in OfflineTts.

mah92 replied:
Shall I add OfflineTtsCacheMechanismConfig to OfflineTtsConfig or keep it separate?

csukuangfj replied:
Please keep it separate.

csukuangfj commented on sherpa-onnx/csrc/offline-tts.h:
```diff
@@ -87,12 +91,26 @@ class OfflineTts {
   // Return the sample rate of the generated audio
   int32_t SampleRate() const;
 
+  // Return the maximum number of cached audio files size
+  int32_t CacheSize() const;
+
+  // Set the maximum number of cached audio files size
+  void SetCacheSize(const int32_t cache_size);
+
+  // Remove all cache data
+  void ClearCache();
+
+  // To get total used cache size(for wav files) in bytes
+  int64_t GetTotalUsedCacheSize();
```
Please remove these methods from OfflineTtts.

You can pass a smart pointer of `CacheMechanism` to the constructor of OffineTts.
(You can add an overload constructor for OfflineTts)

mah92 replied:
I did not get the point.

csukuangfj replied:
We use the `pimpl` implementation here.

csukuangfj commented on sherpa-onnx/csrc/offline-tts.h:
```diff
@@ -87,12 +91,26 @@ class OfflineTts {
   int32_t NumSpeakers() const;
 
  private:
+  OfflineTtsConfig config_;
   std::unique_ptr<OfflineTtsImpl> impl_;
+  std::unique_ptr<CacheMechanism> cache_mechanism_;
```
```suggestion
  std::unique_ptr<OfflineTtsImpl> impl_;
```

mah92 replied:
I did not get how.

mah92 replied:
I do not get it. If CacheMechanism is to be added to the constructor of OffineTts, aren't these lines still needed?

csukuangfj replied:
please update OfflineTtsImpl.

You don't need to add any new fields or any new methods to OfflineTts.

You can add a new field
```
CacheMechanism* cache_; // not owned here
```
to `OfflineTtsImpl`.

csukuangfj commented:
Please use
https://github.com/k2-fsa/sherpa-onnx/blob/master/scripts/check_style_cpplint.sh
to check the style of your C++ code.

You can refer to
https://github.com/k2-fsa/sherpa-onnx/blob/66e02d83815c4f4010e81c479bf690e44b865cae/scripts/check_style_cpplint.sh#L19-L26

---

You can use
```
pip install clang-format==12.0.1
```
to install clang-format and use it to auto-format your code.

mah92 commented:
Thank you, I will correct tomorrow. 
I wonder if it is better to remove the clear button and remove the cache automatically when the speed is changed. By the way the clear button seems to be displaced after merging to the newer commit.

csukuangfj commented:
> I wonder if it is better to remove the clear button and remove the cache automatically when the speed is changed

Yes, I agree.

---

By the way, you can even set a default cache size. Users don't need to set the cache size through the UI.

mah92 commented:
Another proposal: whatever we do in the model, there still remains some words that are not spelled correctly(like keyboard letters). Add them permanently in the cache so that they be read fron the cache instead of the model?

mah92 commented:
About the cache slider, I think it is good to keep it for a while. Then rethink based on user feedback.

csukuangfj commented:
> the cache instead of the model?

Ok, it is fine with me.

csukuangfj commented:
> Another proposal: whatever we do in the model, there still remains some words that are not spelled correctly(like keyboard letters). Good to add them permanently in the cache so that they be read fron the cache instead of the model?

Please make the cache do only 1 thing.

You can create a separate PR to add it.

mah92 commented on android/SherpaOnnxTtsEngine/app/src/main/java/com/k2fsa/sherpa/onnx/tts/engine/MainActivity.kt:
```diff
@@ -88,6 +94,17 @@ class MainActivity : ComponentActivity() {
+                                var usedCacheSizeMB by remember { mutableStateOf(0) }
+
+                                // LaunchedEffect to periodically update the used cache size
+                                LaunchedEffect(Unit) {
```
I searched, LaunchedEffect is indeed a coroutine, and delay is non-blocking.

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.h:
```diff
@@ -0,0 +1,82 @@
+// sherpa-onnx/csrc/offline-tts-cache-mechanism.cc
```
corrected.

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.h:
```diff
@@ -0,0 +1,82 @@
+  void AddWavFile(const std::string &text_hash, const std::vector<float> &samples, int32_t sample_rate);
+
+  // Get the cached wav file if it exists
+  std::vector<float> GetWavFile(const std::string &text_hash, int32_t &sample_rate);
```
Impossible as repeat_counts_ is increased in this function. It is to keep statistics of hashes.

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.h:
```diff
@@ -0,0 +1,82 @@
+  // Data directory where the cache folder is located
+  std::string cache_dir_;
+
+  // Maximum number of files in the cache
```
corrected

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.h:
```diff
@@ -0,0 +1,82 @@
+  void ClearCache();
+
+  // To get total used cache size(for wav files) in bytes
+  int64_t GetTotalUsedCacheSize();
```
corrected

mah92 replied:
Corrected. At the beginning, I tought that the cache might go over 2GB, but now I know that this is not the case.

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+    : cache_dir_(cache_dir), cache_size_bytes_(cache_size), used_cache_size_bytes_(0) {
+  // Create the cache directory if it doesn't exist
+  if (!std::filesystem::exists(cache_dir_)) {
+    std::filesystem::create_directory(cache_dir_);
```
Added cache_mechanism_inited_ variable so that the library can be disabled. Added checking this variable through all public functions.

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+  // Create the cache directory if it doesn't exist
+  if (!std::filesystem::exists(cache_dir_)) {
+    std::filesystem::create_directory(cache_dir_);
+    // SHERPA_ONNX_LOGE("Created cache directory: %s", cache_dir_.c_str());
```
Done

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.h:
```diff
@@ -0,0 +1,82 @@
+  int32_t cache_size_bytes_;
+
+  // Total used cache size for wav files in bytes
+  int64_t used_cache_size_bytes_;
```
done

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+          used_cache_size_bytes_ += file.tellg();
+          file.close();
+        }
+        cache_vector_.push_back(text_hash);
```
done

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+
+void CacheMechanism::EnsureCacheLimit() {
+  if(used_cache_size_bytes_ > cache_size_bytes_) {
+    auto target_cache_size = std::max((int)(cache_size_bytes_*0.95), 0); //Remove more to prevent deleting every step
```
done

mah92 commented on sherpa-onnx/csrc/offline-tts-cache-mechanism.cc:
```diff
@@ -0,0 +1,272 @@
+        std::ifstream file(entry.path(), std::ios::binary | std::ios::ate);
+        if (file.is_open()) {
+          used_cache_size_bytes_ += file.tellg();
+          file.close();
```
done

Commit:
Done some of the reviewer's request

Commit:
cpplint passed

mah92 commented:
> Please use https://github.com/k2-fsa/sherpa-onnx/blob/master/scripts/check_style_cpplint.sh to check the style of your C++ code.
> 
> You can refer to
> 
> https://github.com/k2-fsa/sherpa-onnx/blob/66e02d83815c4f4010e81c479bf690e44b865cae/scripts/check_style_cpplint.sh#L19-L26
> 
> You can use
> 
> ```
> pip install clang-format==12.0.1
> ```
> 
> to install clang-format and use it to auto-format your code.

done.

Commit:
Removed Clear Button, remove cache on speed change

Commit:
Renamed CacheMechanism to OfflineTtsCacheMechanism

Commit:
Switched back to chrono as used in other files

Commit:
Added offline-tts-cache-mechanism-config.cc,.h

mah92 commented:
Shall I add OfflineTtsCacheMechanismConfig to OfflineTtsConfig or keep it separate?

csukuangfj commented:
You don't need to wrap.impl_ to.jni.


You need to.wrap. CacheMechanism to jni.


Not sure why you.need to wrap impl_ to jni.

mah92 commented:
You told me ealier to update OfflineTtsImpl and add CacheMechansim pointer there. It was impossible due to it's private nature. So I should cache the pointer in the  jni/offline-tts.cc?

csukuangfj commented:
Yes, I said that.

Let me give.you the pseudo code later.

Commit:
Using offline-tts-cahce-mechanism-config in both SherpaOnnxTts and SherpaOnnxEngine
Removing cache_dir from OfflineTts
Removing function implementations from OfflineTts
Passing config pointer  to OfflineTts
Added cache_mechanism_ to OfflineTts (as public, might not be so neat)
Changing all cache units in kotlin to Bytes from MB to reduce confusion. Showing is still in MBs

mah92 commented:
This is what I have done in the last commit:

Commit:
Changed all text_hash from string to size_t for more performance.

Also solved bug by accidentally pasted extra code in previous commit

Commit:
Use binary repeat_counts file for better performance, Changed repeat_counts from uint32_t to size_t for better alignment in the file, Changed repeat_counts save time to every 1minute as destructor not working

csukuangfj commented:
So here is the pseudo code.

1. Create a cache mechanism outside of OfflineTts.
```cpp
OfflineTtsCacheMechanismConfig config;
// set config

auto cache = new OfflineTtsCacheMechanism(config)
```

2. Now pass the pointer to `cache` to the OfflineTts
```cpp
OfflineTtsConfig config;

auto tts = new OfflineTts(config, cache);
```

3. Now Change OfflineTtsImpl
```cpp
class OfflineTtsImpl {
public:
  OfflineTtsImpl(const OfflineTtsConfig& config, OfflineTtsCacheMechanism* cache)
   : ...
     cache_(cache) {
   ...
  }

private:
  OfflineTtsCacheMechanism *cache_; // not owned here
};
```

Note again in the above code, you don't need to add any new methods (except the constructor) of OfflineTts.

You also don't need to care about that `impl_` is private in `OfflineTts`.

---

You just need to wrap `OfflineTtsCacheMechanism`  and `OfflineTtsCacheMechanismConfig`. to JNI.

---

Please don't add any method related to OfflineTtsCacheMechanism to `OfflineTts`. You don't need to do that.

csukuangfj commented on sherpa-onnx/csrc/offline-tts.h:
```diff
@@ -91,6 +99,8 @@ class OfflineTts {
   // If it supports only a single speaker, then it return 0 or 1.
   int32_t NumSpeakers() const;
 
+  std::unique_ptr<OfflineTtsCacheMechanism> cache_mechanism_; // not owned here
```
```suggestion
```

Please don't add any new fields to OfflineTts.

As said before, we use `pimpl` here.

If you don't know what `pimpl` is, please google it first.

The only member `OfflineTts` has is `impl_`.

Everything else is put in`OfflineTtsImpl`.

mah92 commented:
I will catch up soon. By the way, What files needs change in the 1. and 2. part of pseudo code change comment?

csukuangfj commented:
> What files needs change in the 1. and 2. part of pseudo code change comment?

> Create a cache mechanism outside of OfflineTts.

You can first wrap the cache mechanism class to kotlin via jni. Then you can create it in kotlin, like how you create offline tts in kotlin.

> Now pass the pointer to cache to the OfflineTts

You need to create offline tts in kotlin, which means you need to pass the instance of cache mechanism from kotlin.

Commit:
bug fix

mah92 committed:
Merge branch 'master' of github.com:k2-fsa/sherpa-onnx into HEAD

mah92 commented:
Hi again. I have progressed, but need some help.
1) "You just need to wrap OfflineTtsCacheMechanism and OfflineTtsCacheMechanismConfig. to JNI."
 What do you mean to wrap them? Would you give me s pseudo code?
Should I cache "cache" as static in jni/offline-tts.cc?
2) I see the warning "This branch must not contain merge commits.". Should I rebase the whole branch?

mah92 committed:
corrections

mah92 committed:
corrections2 - still not compiles

mah92 commented:
By the way, I have made a working thread for cleanup based on user feedback.

mah92 commented:
I think the remaining work is not sth that I can do without your elaborate help. maybe It takes less time from you...

mah92 commented:
After thinking thoroughly and implementing two android apps using sherpa-onnx framework, I think this feature is something the developer should implement in the upper layer, not needed in the sherpa-onnx itself. It makes sherpa too complex and I didn't use this for my apps. I propose to close this. Agree? @csukuangfj

csukuangfj commented:
Yes, please close it.
